harness/harness-core
# [fix]: [DEL-5024]: Perpetual Task Enhancements

## Describe your changes
Summary:
- Remove rebalance iterator and combine re-assigning functionality with  pt assignment iterator
- New delegate disconnect iterator to detect HB expiry and to fail regular and PT tasks on disconnect
- New state "PT_NON_ASSIGNABLE" to track PT's that are not assignable
- https://harness.atlassian.net/wiki/spaces/DEL/pages/21197226916/Perpetual+Task+Enhancements+Q3-22

Test case tested locally

1) Perpetual Task

-  Create new PT and successfully assigned to delegate
-  Create PT assignment fails on no eligible delegate, verify rite state
- Create PT assignment fails on validation, verify rite state
- Verify assign try count is set correctly on ever state change
- Verify assignAfterms set correctly on every state change
- Verify states set correctly on every state change
- verify unassigned reason set correctly on exceptions 
- Verify after 10 assign try count, state change to TASK_NON_ASSIGNABLE


2) Delegate Task
- verify already started task marked as failed on delegate disconnect more than 5 mins


3) Delegate disconnect verification
- Verify delegate disconnect iterator picking up delegate disconnect event
- Verify all delegate tasks assigned to delegate marked as failed 
- Verify all PT's marked as task unassigned state
- Verify delegate disconnection db entry marked as disconnected
- Verify delegate disconnect shows up in UI as soon as delegate disconnected


4) Delegate reconnect verification

- if delegate is reconnecting after 5 minutes, verify all TASK_UNASSIGNABLE pts marked ad TASK_UNASSINED
- Verify all pt's are getting assigned after reconnect


5) New delegate verification
-  when delegate gets added, verify all TASK_UNASSIGNABLE pts marked ad TASK_UNASSINED


6) PT context call fail
-  verify if context call fails, mark PT as Invalid state
- Verify PT running end to end


## Checklist
- [ x] I've documented the changes in the PR description.
- [ x] I've tested this change either in PR or local environment.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/38546)
